crates/derive/src/sources/plasma.rs

```diff
@@ -1,42 +1,177 @@
+            }
+            None => {
+                tracing::warn!("l1 origin advance returned None");
+            }
+        }
+
+        // Set the commitment if it isn't available.
+        if self.commitment.is_none() {
+            // The l1 source returns the input commitment for the batch.
+            let data = match self.source.next().ok_or(PlasmaError::NotEnoughData) {
+                Ok(d) => d,
+                Err(e) => {
+                    tracing::warn!("failed to pull next data from the plasma source iterator");
+                    return Some(Err(StageError::Plasma(e)));
+                }
+            };
+
+            // If the data is empty,
+            if data.is_empty() {
+                return Some(Err(StageError::Plasma(PlasmaError::NotEnoughData)));
+            }
+
+            // If the tx data type is not plasma, we forward it downstream to let the next
+            // steps validate and potentially parse it as L1 DA inputs.
+            if data[0] != TX_DATA_VERSION_1 {
+                return Some(Ok(data));
+            }
+
+            // Validate that the batcher inbox data is a commitment.
+            self.commitment = match decode_keccak256(&data[1..]) {
+                Ok(c) => Some(c),
+                Err(e) => {
+                    tracing::warn!("invalid commitment: {}, err: {}", data, e);
+                    return self.next().await;
+                }
+            };
+        }
+
+        // Use the commitment to fetch the input from the plasma DA provider.
+        let commitment = self.commitment.as_ref().expect("the commitment must be set");
+
+        // Fetch the input data from the plasma DA provider.
+        let data = match self
+            .input_fetcher
+            .get_input(&self.chain_provider, commitment.clone(), self.id)
+            .await
+        {
+            Some(Ok(data)) => data,
+            Some(Err(PlasmaError::ReorgRequired)) => {
+                // The plasma fetcher may call for a reorg if the pipeline is stalled and the plasma
+                // DA manager continued syncing origins detached from the pipeline
+                // origin.
+                tracing::warn!("challenge for a new previously derived commitment expired");
+                return Some(Err(StageError::Reset(ResetError::ReorgRequired)));
+            }
+            Some(Err(PlasmaError::ChallengeExpired)) => {
+                // This commitment was challenged and the challenge expired.
+                tracing::warn!("challenge expired, skipping batch");
+                self.commitment = None;
+                // Skip the input.
+                return self.next().await
+            }
+            Some(Err(PlasmaError::MissingPastWindow)) => {
+                return Some(Err(StageError::Custom(anyhow::anyhow!(
+                    "data for comm {:?} not available",
+                    commitment
+                ))));
```

Comment:
The reference implementation returns a critical error here. We should probably introduce a new Critical `StageError` type to maintain parity with the reference impl.

https://github.com/ethereum-optimism/optimism/blob/6f99544c57ef837afaa96f979a6154a4e8a3604a/op-node/rollup/derive/plasma_data_source.go#L85